lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,453 @@
+        """Return whether credentials have been issued for the request."""
+        relation = self.model.get_relation(self.relation_name, relation_id)
+        if not relation:
+            return False
+        credentials = self.get_credentials(relation)
+        return all(credentials)
+
+    def get_credentials(self, relation: ops.Relation) -> tuple[str | None, str | None]:
+        """Return the credentials from the Juju secret.
+
+        Args:
+            relation: The relation to get the credentials for.
+
+        Returns:
+            A tuple containing the role id and secret id
+        """
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return None, None
+        if relation.app is None:
+            logger.warning("No remote application yet")
+            return None, None
+        credentials_secret_id = relation.data[relation.app].get("credentials_secret_id")
+        if credentials_secret_id is None:
+            return None, None
+        try:
+            secret_content = self.model.get_secret(id=credentials_secret_id).get_content(
+                refresh=True
+            )
+        except SecretNotFoundError:
+            logger.warning("Secret not found")
+            return None, None
+
+        return (secret_content.get("role-id"), secret_content.get("secret-id"))
+
+
+def get_transit_key_name(relation_id: int) -> str:
+    """Return the transit key name for the given relation id.
+
+    Args:
+        relation_id: The id of the relation for which to get the transit key name.
+
+    Returns:
+        The transit key name.
+    """
+    return f"{relation_id}"
+
+
+def is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Return whether the provider data is valid.
+
+    This uses the pydantic schema to validate the data.
+
+    Args:
+        data: The data to validate.
+
+    Returns:
+        True if the data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        data = event.relation.data[event.app]
+        if is_provider_data_valid(data):
+            self.on.vault_autounseal_details_ready.emit(*self.get_details())
```

Comment:
wouldn't `*self.get_details()` make it a bit flakey?
For example if we switch the order of address and ca_certificate in the tuple returned from `get_details` we won't get any error but the user would get the ca_certificate at event.address and the address at event.ca_certificate? or is that not true?

Reply:
That is true. 

That's probably a good argument for following up on namedtuple or a dataclass as mentioned in [this comment](https://github.com/canonical/vault-k8s-operator/pull/346#discussion_r1583634684). I'll change this up.

Reply:
@saltiyazan What are your thoughts on this? https://github.com/canonical/vault-k8s-operator/pull/346/commits/3622bef0686a00cb7270aa6785136343a67eded1

Reply:
Seems more difficult to mess things up and not notice it now